File: android/ui/src/main/java/com/vivianweidai/science/ui/OlympiadsView.kt
package com.vivianweidai.science.ui

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.material3.TopAppBar
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.vivianweidai.science.core.ContentStore
import com.vivianweidai.science.core.grouping.ActivityGrouping
import com.vivianweidai.science.core.model.Activity

/** Unified chronological timeline of olympiads + textbooks, visually
 *  matching the webapp at /olympiads/. Groups by year (newest first). */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun OlympiadsView(
    store: ContentStore,
    modifier: Modifier = Modifier,
) {
    val activities by store.activities.collectAsStateWithLifecycle()
    val error by store.activitiesError.collectAsStateWithLifecycle()
    var filter by rememberSaveable(
        stateSaver = androidx.compose.runtime.saveable.Saver<SubjectFilter, String>(
            save = { (it as? SubjectFilter.Named)?.name ?: "__all__" },
            restore = { s -> if (s == "__all__") SubjectFilter.All else SubjectFilter.Named(s) },
        )
    ) { mutableStateOf<SubjectFilter>(SubjectFilter.randomSubject()) }

    Scaffold(
        modifier = modifier,
        topBar = {
            TopAppBar(
                title = { Text("Olympiads") },
                actions = {
                    SubjectFilterMenu(
                        selected = filter,
                        onSelect = { filter = it },
                    )
                },
            )
        },
    ) { padding ->
        Box(Modifier.padding(padding)) {
            when {
                activities != null -> Timeline(
                    groups = yearGroups(activities!!, filter),
                )
                error != null -> ErrorState(error!!)
                else -> LoadingState(
                    title = "Loading olympiads",
                    subtitle = "Fetching the latest timeline from GitHub.",
                )
            }
        }
    }
}

private fun yearGroups(
    activities: List<Activity>,
    filter: SubjectFilter,
): List<ActivityGrouping.YearGroup> {
    val subjectName = (filter as? SubjectFilter.Named)?.name
    val filtered = ActivityGrouping.filtered(activities, subjectName)
    return ActivityGrouping.groupedByYear(filtered)
}

@Composable
private fun Timeline(
    groups: List<ActivityGrouping.YearGroup>,
) {
    LazyColumn(
        contentPadding = PaddingValues(bottom = 24.dp),
    ) {
        groups.forEach { group ->
            item(key = "year-${group.year}") {
                Text(
                    text = group.year,
                    fontSize = 17.sp,
                    fontWeight = FontWeight.Bold,
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(horizontal = 16.dp)
                        .padding(top = 16.dp, bottom = 4.dp),
                )
            }
            items(group.entries, key = { "act-${it.id}" }) { entry ->
                ActivityRow(entry)
                HorizontalDivider(modifier = Modifier.padding(start = 16.dp))
            }
        }
    }
}

@Composable
private fun ActivityRow(a: Activity) {
    val month = remember(a.date) {
        a.date.substringBefore(' ').take(3)
    }
    val chips = a.subjects?.takeIf { it.isNotEmpty() } ?: listOf(a.subject)
    // Canonical curriculum highlight yellow — #fff056, matching
    // --highlight-bg in archives/android/ui/src/main/assets/katex-shell.html
    // and content/research/curriculum.css. Keep these in sync.
    val highlighted = a.highlighted == 1
    val rowBg = if (highlighted)
        Color(red = 1.0f, green = 0.941f, blue = 0.337f)
    else Color.Transparent
    // Force dark ink on highlighted rows — in dark mode the default
    // body color is near-white, which is unreadable on the yellow wash.
    val nameColor = if (highlighted) Color.Black.copy(alpha = 0.82f)
                    else MaterialTheme.colorScheme.onSurface
    val monthColor = if (highlighted) Color.Black.copy(alpha = 0.55f)
                     else MaterialTheme.colorScheme.onSurfaceVariant

    Row(
        modifier = Modifier
            .fillMaxWidth()
            .background(rowBg)
            .padding(horizontal = 16.dp, vertical = 8.dp),
        verticalAlignment = Alignment.Top,
    ) {
        Text(
            text = month,
            fontSize = 12.sp,
            fontFamily = FontFamily.Monospace,
            color = monthColor,
            modifier = Modifier.width(56.dp),
        )
        Box(Modifier.width(24.dp).padding(end = 4.dp), contentAlignment = Alignment.Center) {
            if (a.isOlympiad) OlympicRings() else Text("📖", fontSize = 13.sp)
        }
        Column(Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(4.dp)) {
            Row(horizontalArrangement = Arrangement.spacedBy(4.dp)) {
                chips.forEach { SubjectChip(it) }
            }
            Row(verticalAlignment = Alignment.Top) {
                Text(
                    text = a.name,
                    fontSize = 14.sp,
                    color = nameColor,
                    modifier = Modifier.weight(1f, fill = false),
                )
                if (a.invited == 1) Text(" ⭐", fontSize = 12.sp)
                if (a.competitive == 1) Text(" 🎯", fontSize = 12.sp)
            }
        }
    }
}

/** Simplified Olympic rings — 5 circles in two rows matching the webapp SVG. */
@Composable
private fun OlympicRings() {
    Canvas(modifier = Modifier.size(width = 22.dp, height = 12.dp)) {
        val r = size.minDimension * 0.29f
        val stroke = 1.2.dp.toPx()
        fun drawRing(color: Color, cx: Float, cy: Float) {
            drawCircle(color, radius = r, center = androidx.compose.ui.geometry.Offset(cx, cy),
                style = androidx.compose.ui.graphics.drawscope.Stroke(width = stroke))
        }
        val midY = size.height * 0.42f
        val lowY = size.height * 0.66f
        drawRing(Color(0.00f, 0.51f, 0.78f), size.width * 0.18f, midY) // blue
        drawRing(Color.Black,                 size.width * 0.50f, midY)
        drawRing(Color(0.93f, 0.20f, 0.31f), size.width * 0.82f, midY) // red
        drawRing(Color(0.99f, 0.69f, 0.19f), size.width * 0.34f, lowY) // yellow
        drawRing(Color(0.00f, 0.65f, 0.32f), size.width * 0.66f, lowY) // green
    }
}

File: apple/shared/UI/Views/OlympiadsView.swift
import SwiftUI
import ScienceCore

/// Unified chronological timeline of olympiads + textbooks, visually matching
/// the webapp at /olympiads/. Groups entries by year (newest first), renders
/// each row with a month label, type icon, subject chips, and name with
/// optional INVITED badge.
///
/// The webapp's implementation lives in olympiads/index.md — keep this view
/// in sync with the timeline JS there when the design changes.
struct OlympiadsView: View {
    @State private var store = ContentStore.shared
    @State private var subject: SubjectFilter = SubjectFilter.randomSubject()

    var body: some View {
        NavigationStack {
            Group {
                if let activities = store.activities {
                    timelineContent(activities: activities)
                } else if let errorMessage = store.activitiesError {
                    ErrorState(message: errorMessage)
                } else {
                    LoadingState(
                        title: "Loading olympiads",
                        subtitle: "Fetching the latest timeline from GitHub."
                    )
                }
            }
            .navigationTitle("Olympiads")
            .toolbar {
                ToolbarItem(placement: .primaryAction) {
                    SubjectFilterMenu(selected: $subject)
                }
            }
            .refreshable { await store.refreshAll() }
        }
    }

    // MARK: - Timeline

    private func timelineContent(activities: [Activity]) -> some View {
        ScrollView {
            LazyVStack(alignment: .leading, spacing: 0) {
                ForEach(yearGroups(for: activities), id: \.year) { group in
                    YearSection(year: group.year, entries: group.entries)
                }
            }
            .padding(.bottom, 24)
        }
    }

    // Grouping semantics shared with watchOS via `ScienceCore`.
    private func yearGroups(for activities: [Activity]) -> [ActivityGrouping.YearGroup] {
        let subjectName: String? = {
            if case .named(let name) = subject { return name }
            return nil
        }()
        let filtered = ActivityGrouping.filtered(activities, subject: subjectName)
        return ActivityGrouping.groupedByYear(filtered)
    }
}

// MARK: - Subject filter

private enum SubjectFilter: Hashable {
    case all
    case named(String)

    static let allCases: [SubjectFilter] = [
        .all,
        .named("Mathematics"),
        .named("Computing"),
        .named("Physics"),
        .named("Chemistry"),
        .named("Biology"),
        .named("Astronomy"),
    ]

    /// Pick a random non-"All" subject for the initial filter on launch,
    /// matching the webapp which randomly preselects one of the six subject
    /// tabs each page load (see olympiads/index.md line 16-18).
    static func randomSubject() -> SubjectFilter {
        let subjects: [SubjectFilter] = allCases.filter {
            if case .all = $0 { return false }
            return true
        }
        return subjects.randomElement() ?? .all
    }

    var label: String {
        switch self {
        case .all: return "All"
        case .named(let n): return n
        }
    }

    var color: Color? {
        switch self {
        case .all: return nil
        case .named(let n): return SubjectPalette.color(for: n)
        }
    }
}

/// Native toolbar filter: a "Filter" button with a colored dot for the
/// active subject opens a system menu of all subjects with a checkmark on
/// the current selection. Adapts to light/dark mode automatically.
private struct SubjectFilterMenu: View {
    @Binding var selected: SubjectFilter

    var body: some View {
        Menu {
            // Plain Button rows (not a Picker) so each subject dot can
            // carry its own palette color. SwiftUI's Picker inside a
            // Menu renders system-tinted symbols only.
            ForEach(SubjectFilter.allCases, id: \.self) { filter in
                Button { selected = filter } label: {
                    filterMenuRow(filter: filter, isSelected: filter == selected)
                }
            }
        } label: {
            HStack(spacing: 4) {
                if let color = selected.color {
                    Circle().fill(color).frame(width: 10, height: 10)
                }
                Text(selected.label)
                    .font(.system(size: 15, weight: .medium))
            }
        }
    }
}

/// Menu row for Olympiads subject filter. ResearchView carries its own
/// near-identical copy because `SubjectFilter` is file-private to each
/// view.
@ViewBuilder
private func filterMenuRow(filter: SubjectFilter, isSelected: Bool) -> some View {
    HStack {
        Text(filter.label)
        Spacer()
        if isSelected {
            Image(systemName: "checkmark")
        }
        if let color = filter.color {
            Circle().fill(color).frame(width: 12, height: 12)
        } else {
            Image(systemName: "square.grid.2x2")
        }
    }
}

/// Initial-load placeholder shown while the first network fetch is in
/// flight. A bare `ProgressView()` reads as a blank screen on iPhone;
/// this wraps it with a title + subtle subtitle so the wait feels
/// intentional. Used by all three top-level tabs.
struct LoadingState: View {
    let title: String
    let subtitle: String

    var body: some View {
        VStack(spacing: 14) {
            ProgressView()
                .controlSize(.large)
            Text(title)
                .font(.system(size: 15, weight: .semibold))
                .foregroundStyle(.primary)
            Text(subtitle)
                .font(.system(size: 12))
                .foregroundStyle(.secondary)
                .multilineTextAlignment(.center)
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
    }
}

/// Simple error display used when the initial fetch fails. Avoids
/// `ContentUnavailableView` which requires macOS 14 and complicates
/// host type-checking via `swift build`.
struct ErrorState: View {
    let message: String

    var body: some View {
        VStack(spacing: 10) {
            Image(systemName: "exclamationmark.triangle")
                .font(.system(size: 28))
                .foregroundStyle(.secondary)
            Text("Could not load")
                .font(.headline)
            Text(message)
                .font(.caption)
                .foregroundStyle(.secondary)
                .multilineTextAlignment(.center)
        }
        .padding()
    }
}

// MARK: - Year section

private struct YearSection: View {
    let year: String
    let entries: [Activity]

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            Text(year)
                .font(.system(size: 17, weight: .bold))
                .padding(.horizontal)
                .padding(.top, 16)
                .padding(.bottom, 4)

            ForEach(entries) { entry in
                ActivityRow(activity: entry)
                Divider()
                    .padding(.leading, 16)
            }
        }
    }
}

// MARK: - Activity row

private struct ActivityRow: View {
    let activity: Activity

    /// The month label shown in the timeline column. The source-of-truth
    /// dates are full month names ("December 2025"); we abbreviate to three
    /// letters here so longer names don't wrap in the narrow fixed-width
    /// column. Source data is untouched.
    private var month: String {
        let full = activity.date.split(separator: " ").first.map(String.init) ?? ""
        return String(full.prefix(3))
    }

    private var chips: [String] {
        if let subjects = activity.subjects, !subjects.isEmpty {
            return subjects
        }
        return [activity.subject]
    }

    var body: some View {
        let highlighted = activity.highlighted == 1
        // Force dark ink on highlighted rows — in dark mode the default
        // body color is near-white, which is unreadable on the yellow wash.
        // Matches Android OlympiadsView.kt.
        let nameColor: Color = highlighted ? Color.black.opacity(0.82) : Color.primary
        let monthColor: Color = highlighted ? Color.black.opacity(0.55) : Color.secondary

        return HStack(alignment: .firstTextBaseline, spacing: 8) {
            Text(month)
                .font(.system(size: 12, weight: .regular, design: .monospaced))
                .foregroundStyle(monthColor)
                .frame(width: 56, alignment: .leading)

            TypeIcon(isOlympiad: activity.isOlympiad)
                .frame(width: 24)

            VStack(alignment: .leading, spacing: 4) {
                HStack(spacing: 4) {
                    ForEach(chips, id: \.self) { subject in
                        SubjectChip(subject: subject)
                    }
                }
                HStack(alignment: .firstTextBaseline, spacing: 4) {
                    Text(activity.name)
                        .font(.system(size: 14))
                        .foregroundStyle(nameColor)
                        .fixedSize(horizontal: false, vertical: true)
                    if activity.invited == 1 || activity.attended == 1 {
                        Text("⭐").font(.system(size: 12))
                    }
                    if activity.competitive == 1 {
                        Text("🎯").font(.system(size: 12))
                    }
                    if activity.alternate == 1 {
                        Text("🇨🇦").font(.system(size: 12))
                    }
                }
            }

            Spacer(minLength: 0)
        }
        .padding(.horizontal, 16)
        .padding(.vertical, 8)
        // Canonical curriculum highlight yellow — #fff056, matching
        // --highlight-bg in archives/apple/shared/UI/Rendering/katex-shell.html
        // and content/research/curriculum.css. Keep these in sync.
        .background(
            highlighted
                ? Color(red: 1.0, green: 0.941, blue: 0.337)
                : Color.clear
        )
    }
}

// MARK: - Subject chip

private struct SubjectChip: View {
    let subject: String

    var body: some View {
        Text(subject)
            .font(.system(size: 10, weight: .semibold))
            .padding(.horizontal, 7)
            .padding(.vertical, 2)
            .background(Capsule().fill(SubjectPalette.color(for: subject)))
            .foregroundStyle(Color.black.opacity(0.82))
    }
}

// MARK: - Type icons

private struct TypeIcon: View {
    let isOlympiad: Bool

    var body: some View {
        if isOlympiad {
            OlympicRings()
                .frame(width: 22, height: 12)
        } else {
            Text("📖")
                .font(.system(size: 13))
        }
    }
}

/// Simplified Olympic rings — 5 circles in two rows matching the webapp SVG.
private struct OlympicRings: View {
    var body: some View {
        ZStack {
            ring(color: Color(red: 0.0, green: 0.51, blue: 0.78), offset: CGPoint(x: -8, y: -2))    // blue
            ring(color: .black, offset: CGPoint(x: 0, y: -2))
            ring(color: Color(red: 0.93, green: 0.20, blue: 0.31), offset: CGPoint(x: 8, y: -2))   // red
            ring(color: Color(red: 0.99, green: 0.69, blue: 0.19), offset: CGPoint(x: -4, y: 3))   // yellow
            ring(color: Color(red: 0.0, green: 0.65, blue: 0.32), offset: CGPoint(x: 4, y: 3))     // green
        }
    }

    private func ring(color: Color, offset: CGPoint) -> some View {
        Circle()
            .strokeBorder(color, lineWidth: 1.2)
            .frame(width: 7, height: 7)
            .offset(x: offset.x, y: offset.y)
    }
}

// MARK: - Subject palette

/// Thin SwiftUI wrapper over `ScienceCore.SubjectPaletteRGB`. The raw
/// (r,g,b) tuples live in Core so the watchOS companion can share them
/// without dragging SwiftUI into the platform-neutral target.
enum SubjectPalette {
    static func color(for subject: String) -> Color {
        let rgb = SubjectPaletteRGB.rgb(for: subject)
        return Color(red: rgb.red, green: rgb.green, blue: rgb.blue)
    }
}
